oeps/oep-0045/decisions/0001-tutor-as-replacement-for-edx-configuration.rst

```diff
@@ -0,0 +1,328 @@
+
+And via `officially supported plugins <https://docs.tutor.overhang.io/plugins.html#existing-plugins>`__, one can easily
+add these other official Open edX components:
+
+-  Course Discovery
+-  Ecommerce
+-  Notes
+-  Xqueue
+
+Notably absent is any support for deployment of MFEs. Given that not even ``edx/configuration`` currently supports this
+(though there is `a PR <https://github.com/edx/configuration/pull/6128>`__ for it, along with `instructions
+<https://discuss.openedx.org/t/deploying-mfes-in-the-community/2868/22>`__), this shouldn't count against it. In any
+case, the project leadership is `aware of the issue
+<https://discuss.overhang.io/t/work-in-progress-upgrading-tutor-to-koa/1071>`__.
+
+
+External databases
+``````````````````
+
+An important characteristic of scaling up an Open edX deployment is having the ability to point it to external
+databases, such as Amazon RDS or a manually maintained Galera cluster. Tutor allows for just this. All of the following
+can be disabled locally and easily `configured to use an externally hosted service
+<https://docs.tutor.overhang.io/configuration.html#mysql>`__:
+
+-  MySQL
+-  Elasticsearch
+-  MongoDB
+-  Redis
+-  SMTP
+
+
+Multi-server
+````````````
+
+Single-server deployments are less than ideal in terms of high availability and scalability, so for larger deployments
+something different is needed. And given that Tutor deployments use Docker images, Kubernetes would be the first choice.
+(Not least of which because Kubernetes is also what edX `has started to use
+<https://discuss.openedx.org/t/lets-talk-about-the-native-installation/3269/14>`__.)
+
+Tutor does come with basic support for Kubernetes deployment.  ``tutor k8s quickstart`` on a local `MicroK8s
+<https://microk8s.io/>`__ 1.8 installation did not work out of the box, though. There is a race condition during the
+startup of the MySQL container that requires relaunching the command, and one has to modify the MinIO hostname manually
+to avoid a failure during the initial LMS job. There were further MinIO name resolutions issues that precluded a fully
+working environment: there was no time to ascertain whether the cause is ultimately with Tutor or with the tested
+microk8s setup.
+
+It is mentioned at the top of the `Tutor Kubernetes documentation <https://docs.tutor.overhang.io/k8s.html>`__ that "the
+local Open edX install was designed such that users with no prior experience with system administration could still
+launch an Open edX platform. It is not the case for the installation method outlined here." While the Tutor project has
+clearly expended effort in getting it to work with Kubernetes, if it is to fully replace ``edx/configuration`` as the
+go-to community installation for production use, it is suggested that the Kubernetes documentation be extended with more
+detailed information on how to run and maintain Open edX in such an environment, and that Kubernetes support be made a
+first-class citizen.
```

Comment:
This is a fairly controversial statement :) Let me argue a few points:

1. Yes, the Tutor docs on K8s deployments are quite sparse and should be improved. However...
2. ... Kubernetes deployment is a *very* complex topic in itself -- at least as complex as Open edX, for comparison. K8s versions and implementations vary from one provider to the next, and it is not Tutor's job to resolve every K8s deployment issue. K8s deployments require some level of expertise and the Tutor docs should not attempt to make up for the lack of that expertise. Note that I am writing this as a Kubernetes noob myself ;) I think that `tutor k8s` should efficiently work with one single implementation of K8s (so far I ran my tests mostly with DigitalOcean) and that it should assumed that all others require some level of K8s and Open edX expertise.
3. I am not sure at all that Kubernetes deployment should be our primary deployment target. Actually, let me rephrase that: Kubernetes should _not_ be our primary deployment target. For the following reasons:

    a. I'd wage that more than 95% of Tutor users don't ever run Open edX on Kubernetes and are happy with docker-compose. Even teams with large Open edX platforms may not be interested by Kubernetes because they lack the required skills, internally.

    b. [There are alternatives to Kubernetes](https://regisb.github.io/cncf2020/#/27): [Docker Swarm](https://docs.docker.com/engine/swarm/) and [Nomad](nomadproject.io/) come to mind. These are _much_ simpler and support multi-server deployments, which is our primary objective.

4. "High availability" is much more complex than simply adding servers when the load increases. In particular, what does high availability mean for the data storage layer? Scaling MySQL, MongoDb, Elasticsearch, Redis and MinIO is very much out of the scope of Tutor. Same for the load balancer. Tutor provides the instrumentation for measuring container load (with logs and container observability), and it is the operator's responsibility to scale their platform according to their requirements and capabilities.

In a nutshell, Tutor is as much what it provides and what it doesn't: if an operator wants to run an auto-scalable Open edX platform with Kubernetes or any other orchestration solution that follows the best practices, then Tutor really should not stand in their way. But it is not Tutor's role to fulfill each and every requirement.

Reply:
@regisb, I'm with @antoviaque on this.  There definitely should be some support, even if just in the form of reference documentation, in how to run an Open edX deployment at the large scale end of the spectrum.  I also agree that it doesn't need to be the main target - but just like `edx/configuration` provides examples on how large-scale is done but doesn't do so by default, the same thing should happen to Tutor if it is to replace it.

Let me put it a different way: where _would_ such support/documentation live, if not as part of the community's official orchestration tool?

As I suggest above, Helm charts in a `contrib/` directory would be one way to go about it.  Maybe we can even convince someone at edX to help out, since they're already using it.  (@fredsmith?)

Reply:
@regisb What can we agree on for this? The support for larger setups is an important element to sort out to take a decision, as it's going to be a frequent use case for many of the Open edX operators currently considering the potential switch to Tutor. I know it has been a secondary concern for Tutor so far, as most of the current audience is more interested in getting a small setup up and running, but making Tutor the official installation path does extend its scope.

Reply:
I'm not sure what the community needs in addition to what already exists -- maybe some better documentation? To clarify: it's important to me that Tutor supports scalable deploys -- and _this is already the case_. To do so:

1. Install and enable the [minio](https://github.com/overhangio/tutor-minio) plugin
2. Deploy to Kubernetes
3. Scale your LMS, CMS and forum containers as much as you want.

It's really as simple as that, and many Tutor users already follow this pattern.

Reply:
@regisb That's great to hear! It sounds like it should be possible to get to something that works for everyone then.

@arbrandes As discussed yesterday during the [contributors meetup](https://discuss.openedx.org/t/open-edx-contributors-meetup/1450/40), could you plan to test this, along with your investigation of the current solution being used by edX for their deployments?